### PR TITLE
[proxy] Update marketplace to use a proxy

### PR DESCRIPTION
**Pull request by awgreene**
This commit introduces a change to enable marketplace to use a globally
configured proxy for its HTTP and HTTPS requests. To enable this
feature, the following changes were made:
* When the marketplace operator starts, a check will be made to
  identify if the Proxy API is available. If the API is available,
  marketplace will watch for changes to the "cluster" proxy. The
  cluster proxy is used to provide global proxy information.
* The CatalogSourceConfig and OperatorSource controllers have had a
  watch added that will alert them when a proxy is created, deleted,
  or destroyed. When the event comes from the cluster proxy,
  the controller will check if any child Resource deployments are not
  in sync with the environment variables defined by the cluster proxy.
* The marketplace operator will be configured to reach out to quay via
  the proxy.

### CODE REVIEW

File: pkg/proxy/envvar.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+// Remove removes the given index from the list
+func Remove(list []corev1.EnvVar, index int) []corev1.EnvVar {
+	return append(list[:index], list[index+1:]...)
+}
+
+// Manage will ensure that the list of EnvVars are in sync with the proxy variables.
+// If the proxy variables are not set but are present in the list, they will be removed from the list.
+// If the proxy variables are set but missing from the list, they will be added to the list.
+// If the proxy variables are set and present in the list, their values will be updated in the list.
+func Manage(list []corev1.EnvVar) []corev1.EnvVar {
```
Comment (aravindhp) [inline pkg/proxy/envvar.go]: I understand this was introduced so that you can ensure that you don't remove any user added environment variables plus you will only add the proxy variables only when required.. I am not sure if we want the user to be able to add their own environment variables given the `operator-registry` application is what is run by the deployment. Also given the list management you are doing here it might be preferable to always have the proxy environment variables in the spec even when they are empty.

In other words this could just be `SetEnvVars()`

File: pkg/proxy/helpers.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+var (
+	// ClusterProxyKey is the key for the cluster proxy.
+	ClusterProxyKey = types.NamespacedName{Name: ClusterProxyName}
+)
+
+// NeedsUpdate return true if the deployment is out of sync with the proxy environment variables.
+func NeedsUpdate(registryDeployment apps.Deployment) bool {
+	// Get the list of environment variables defined in the registry deployment
+	actualEnvVars := registryDeployment.Spec.Template.Spec.Containers[0].Env
```
Comment (aravindhp) [inline pkg/proxy/helpers.go]: Please check if the list is empty and return true if it is empty.

File: pkg/registry/registry.go [hunk 3]
```diff
@@ -137,6 +138,9 @@ func (r *registry) ensureDeployment(appRegistries []string, needServiceAccount b
 			if needServiceAccount {
 				deployment.Spec.Template.Spec.ServiceAccountName = r.key.Name
 			}
+			// Add, Update, or Remove the proxy env vars based on their values in the marketplace pod.
+			deployment.Spec.Template.Spec.Containers[0].Env = proxy.Manage(deployment.Spec.Template.Spec.Containers[0].Env)
```
Comment (aravindhp) [inline pkg/registry/registry.go]: This would become `SetEnvVars()`

File: pkg/watches/helpers.go [hunk 4]
```diff
@@ -171,3 +175,75 @@ func getCscOwnerKey(labels map[string]string) *client.ObjectKey {
...
+	pred := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			if e.Meta.GetName() == proxy.ClusterProxyName {
+				return true
+			}
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// If the cluster operator is ever changed we should kick off an event.
+			if e.MetaOld.GetName() == proxy.ClusterProxyName || e.MetaNew.GetName() == proxy.ClusterProxyName {
```
Comment (aravindhp) [inline pkg/watches/helpers.go]: This check implies that the name changed which is not possible for kube resources. Checking either should be sufficient.

File: pkg/watches/helpers.go [hunk 5]
```diff
@@ -171,3 +175,75 @@ func getCscOwnerKey(labels map[string]string) *client.ObjectKey {
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CheckProxyResource returns true if the deployment with the given namespacedname
+// needs to have its environment variables update due to a change in the cluster proxy.
+func CheckProxyResource(client cl.Client, name, namespace string) (bool, error) {
```
Comment (aravindhp) [inline pkg/watches/helpers.go]: This is a proxy helper and should live in the proxy package.

File: pkg/watches/proxy_csc_mapper.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+type proxyToCatalogSourceConfigs struct {
+	client client.Client
+}
+
+// Map will update the proxy environment variables and create a reconcile request for
+// all CatalogSourceConfig.
+func (m *proxyToCatalogSourceConfigs) Map(obj handler.MapObject) []reconcile.Request {
+	// Ensure that proxy environment variables are in sync.
```
Comment (aravindhp) [inline pkg/watches/proxy_csc_mapper.go]: Will help if you mention this is for the operator and you are using that to track if changes are required for the individual requests.

File: test/e2e/main_test.go [hunk 7]
```diff
@@ -17,6 +17,7 @@ import (
 )
 
 var coAPIPresent = true
+var ProxyAPIPresent = true
```
Comment (aravindhp) [inline test/e2e/main_test.go]: Why do you need this variable?

File: test/e2e/main_test.go [hunk 8]
```diff
@@ -29,6 +30,8 @@ func TestMarketplace(t *testing.T) {
 	// Run Test Groups
 	if coAPIPresent {
 		t.Run("cluster-operator-status-test-group", testgroups.ClusterOperatorTestGroup)
+		// If CO API is available, assume proxy api is as well.
```
Comment (aravindhp) [inline test/e2e/main_test.go]: Then do you really need this variable. You might as well treat `coAPIPresent` as `onOpenShiftCluster`

File: test/testsuites/proxytests.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// ProxyTests is a test suite that ensure that the watches for proxy resources
+// are firing correctly and the deployment resources are updated as a result.
+func ProxyTests(t *testing.T) {
+	t.Run("no-proxy", testSetNoProxyVariable)
+}
+
+// testSetNoProxyVariable creates an OperatorSource and then set the cluster proxy.Spec.NoProxy
```
Comment (aravindhp) [inline test/testsuites/proxytests.go]: Where are you creating the `OperatorSource` in this test case?

File: test/testsuites/proxytests.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+		} else {
+			assert.Equal(t, v.Value, "")
+		}
+	}
+
+	// Remove the NoProxy value in the proxy spec.
+	clusterProxy.Spec.NoProxy = ""
+	err = client.Update(context.TODO(), clusterProxy)
+	require.NoError(t, err, "Could not update the clusterProxy")
+}
```
Comment (aravindhp) [inline test/testsuites/proxytests.go]: For sake of completeness you might want to do the same test for `CatalogSourceConfigs` as you are adding watches there too. Lot of this code can be shared between the two test cases similar to what I did with watches for child resources.

File: deploy/upstream/05_role.yaml [hunk 11]
```diff
@@ -3,6 +3,14 @@ apiVersion: rbac.authorization.k8s.io/v1
 metadata:
   name: marketplace-operator
 rules:
+- apiGroups:
+  - config.openshift.io
```
[the remark this answers is not in the document]
Comment (awgreene) [reply deploy/upstream/05_role.yaml]: This is a good catch.

File: cmd/manager/main.go [hunk 12]
```diff
@@ -84,6 +98,11 @@ func main() {
 		exit(err)
 	}
 
+	// Add proxy resource to scheme
+	if err := apiconfigv1.AddToScheme(mgr.GetScheme()); err != nil {
+		exit(err)
```
[the remark this answers is not in the document]
Comment (awgreene) [reply cmd/manager/main.go]: This does not fail on Kube, I have tested it.

File: cmd/manager/main.go [hunk 13]
```diff
@@ -59,6 +62,17 @@ func main() {
...
 
+	// Set proxy API availability
+	k8sInterface, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = proxy.SetProxyAvailability(k8sInterface.Discovery())
+	if err != nil {
+		log.Fatal(err)
```
[the remark this answers is not in the document]
Comment (awgreene) [reply cmd/manager/main.go]: This method will only attempt to set proxy availability if the Proxy API exists on the cluster. Since CVO is managing the cluster proxy and marketplace, the operator will eventually resolve.

File: pkg/appregistry/appregistry.go [hunk 14]
```diff
@@ -40,7 +42,17 @@ func (f *factory) New(options Options) (Client, error) {
 		return nil, err
 	}
 
-	transport := httptransport.New(u.Host, u.Path, []string{u.Scheme})
+	// A custom HTTPClient is used here since the default HTTPClients ProxyFromEnvironment
+	// uses a cache which won't let us update the proxy env vars.
+	cli := &http.Client{
+		Transport: &http.Transport{
+			Proxy: func(req *http.Request) (*url.URL, error) {
+				return httpproxy.FromEnvironment().ProxyFunc()(req.URL)
```
[the remark this answers is not in the document]
Comment (awgreene) [reply pkg/appregistry/appregistry.go]: This will work if the proxy variables are not set in the environment.

File: cmd/manager/main.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply cmd/manager/main.go]: if you can't get a config it makes sense to me to error hard, we wouldn't be able to contact k8s and I bet that `manager.New` would also fail and you would log.Fatal on line 82. 

The question for setting the proxy availability, I think is more nuanced, what are the error scenarios for this and does it make sense, that if something were to go wrong you would want the operator to continue.

File: pkg/appregistry/appregistry.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (awgreene) [reply pkg/appregistry/appregistry.go]: This does not return an error, we are setting the Proxy func: https://github.com/golang/go/blob/master/src/net/http/transport.go#L112-L121

File: pkg/catalogsourceconfig/succeeded.go [hunk 15]
```diff
@@ -56,9 +61,23 @@ func (r *succeededReconciler) Reconcile(ctx context.Context, in *v2.CatalogSourc
...
 	}
 
-	r.log.Info(msg)
+	if needsProxyUpdate {
+		// The Environment Variables in the deployment created by the CatalogSourceConfig
+		// are out of sync with those in the Cluster Proxy.
+		// Drop the existing Status field so that reconciliation can start anew.
+		out.Status = v2.CatalogSourceConfigStatus{}
+		nextPhase = phase.GetNext(phase.Configuring)
+		msg = "Proxy environment variables not in sync, scheduling for configuring"
```
[the remark this answers is not in the document]
Comment (awgreene) [reply pkg/catalogsourceconfig/succeeded.go]: That is a great idea and will be implemented in the next update to the PR. Perhaps we could add it to the `watches` package so it can be shared between the `CSC` and `opsrc` packages? Otherwise I will make a function in this file for each.

File: pkg/appregistry/appregistry.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply pkg/appregistry/appregistry.go]: https://godoc.org/net/http#ProxyFromEnvironment

You could use that and make it more clear.

Also, this is how the default client works, so this might not even be needed

File: pkg/proxy/available.go [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+// SetProxyAvailability will set isAPIAvailable to the correct value if
+// no unexpected errors are encountered.
+func SetProxyAvailability(discovery apidiscovery.DiscoveryInterface) error {
+	if discovery == nil {
+		return errors.New("discovery interface can not be <nil>")
+	}
+
+	opStatusGV := schema.GroupVersion{
+		Group:   "config.openshift.io",
```
[the remark this answers is not in the document]
Comment (awgreene) [reply pkg/proxy/available.go]: Good idea, I will use the `GroupName` and `GroupVersion.Version` defined here: https://github.com/openshift/api/blob/master/config/v1/register.go#L10-L11

File: pkg/proxy/envvar.go [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"os"
+
+	apiconfigv1 "github.com/openshift/api/config/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	// HTTP_PROXY is the URL of the proxy for HTTP requests.
+	envHTTPProxyName = "HTTP_PROXY"
```
[the remark this answers is not in the document]
Comment (awgreene) [reply pkg/proxy/envvar.go]: Understood, I will update the variables accordingly.

File: pkg/watches/helpers.go (hunk 4 above)
Comment (awgreene) [reply pkg/watches/helpers.go]: This is good to know, thank you Aravindh.

File: test/testsuites/proxytests.go (hunk 9 above)
Comment (awgreene) [reply test/testsuites/proxytests.go]: This test is ran as part of the `OperatorSource` TestGroup, which creates an `OperatorSource` named `test-operators` prior to running this test.

File: test/testsuites/proxytests.go (hunk 10 above)
Comment (awgreene) [reply test/testsuites/proxytests.go]: Sure thing!

File: test/e2e/main_test.go (hunk 7 above)
Comment (awgreene) [reply test/e2e/main_test.go]: Good point - I had meant to remove this. I wanted to run the proxy tests part of the `OperatorGroup` tests but ran into a circular dependency issue. I instead exported a function in the `testgroups` package that lets you set a variable to run the proxy tests.

File: test/e2e/main_test.go (hunk 8 above)
Comment (awgreene) [reply test/e2e/main_test.go]: This will still be needed, please read above.

Comment (awgreene) [reply test/e2e/main_test.go]: Perhaps naming the function `EnableProxyTests` would better portray the point of this function.

File: cmd/manager/main.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (aravindhp) [reply cmd/manager/main.go]: I am a little puzzled by this. Moreover the logic should be that if the proxy API is not present we should not even be attempting to add it to the schem.

File: pkg/catalogsourceconfig/succeeded.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (aravindhp) [reply pkg/catalogsourceconfig/succeeded.go]: This specific to `CatalogSourceConfigs` and `OperatorSources`. I would suggest adding it to `*_types.go` similar to `EnsureGVK()` etc

File: test/e2e/main_test.go (hunk 8 above)
Comment (aravindhp) [reply test/e2e/main_test.go]: I am asking if you need the `proxyAPIAvailable` variable.

File: test/testsuites/proxytests.go (hunk 9 above)
Comment (aravindhp) [reply test/testsuites/proxytests.go]: I know. You should remove the statement that you are creating the `OperatorSource` in this function.

Comment (awgreene) [reply test/testsuites/proxytests.go]: Ahh

File: cmd/manager/main.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (awgreene) [reply cmd/manager/main.go]: I can move this within an if statement that checks for Proxy API.

File: test/e2e/main_test.go (hunk 8 above)
Comment (awgreene) [reply test/e2e/main_test.go]: We do, because the `coAPIPresent` in the `e2e` package cannot be referenced from the `testgroups` package.

File: pkg/appregistry/appregistry.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (awgreene) [reply pkg/appregistry/appregistry.go]: Thanks @shawn-hurley, I tried using `ProxyFromEnvironment` but the result seems to be cached, setting the env variables did cause marketplace to use the proxy.

File: pkg/apis/operators/v2/catalogsourceconfig_types.go [hunk 18]
```diff
@@ -79,6 +79,12 @@ func (csc *CatalogSourceConfig) EnsureGVK() {
 	csc.SetGroupVersionKind(gvk)
 }
 
+// ForceReconcile forces the CatalogSourceConfig object to be reconciled.
+func (csc *CatalogSourceConfig) ForceReconcile() {
```
Comment (aravindhp) [inline pkg/apis/operators/v2/catalogsourceconfig_types.go]: Thinking a bit for about this, I suggest `ForceUpdate()`.

File: pkg/proxy/helpers.go [hunk 19]
```diff
@@ -12,26 +15,50 @@ const (
 
 var (
 	// ClusterProxyKey is the key for the cluster proxy.
-	ClusterProxyKey = types.NamespacedName{Name: ClusterProxyName}
+	clusterProxyKey = types.NamespacedName{Name: ClusterProxyName}
 )
 
+// CheckProxyResource returns true if the deployment with the given namespacedname
```
Comment (aravindhp) [inline pkg/proxy/helpers.go]: Nit: Did you mean to input the `namespaced` name here? If you not you might want to fix your comment.

File: test/testgroups/opsrccsctests.go [hunk 20]
```diff
@@ -50,12 +50,12 @@ func OpSrcCscTestGroup(t *testing.T) {
...
-	if proxyAPIAvailable {
+	if proxyTestsEnabled {
 		t.Run("proxy-test-suites", testsuites.ProxyTests)
 	}
-	t.Run("opsrc-creation-test-suite", testsuites.OpSrcCreation)
-	t.Run("csc-target-namespace-test-suite", testsuites.CscTargetNamespace)
-	t.Run("packages-test-suite", testsuites.PackageTests)
-	t.Run("csc-invalid-tests", testsuites.CscInvalid)
-	t.Run("watch-tests", testsuites.WatchTests)
+	// t.Run("opsrc-creation-test-suite", testsuites.OpSrcCreation)
```
Comment (aravindhp) [inline test/testgroups/opsrccsctests.go]: I am assuming you missed reverting this before pushing :-)

File: pkg/proxy/envvar.go [hunk 21]
```diff
@@ -1,88 +1,60 @@
...
 
 const (
-	// HTTP_PROXY is the URL of the proxy for HTTP requests.
-	envHTTPProxyName = "HTTP_PROXY"
+	// HTTPProxy is the name of the environment variable that sets the proxy for HTTP requests.
+	HTTPProxy = "HTTP_PROXY"
 
-	// HTTPS_PROXY is the URL of the proxy for HTTPS requests.
-	envHTTPSProxyName = "HTTPS_PROXY"
+	// HTTPSProxy is the name of the environment variable that sets the proxy for  HTTPS requests.
```
Comment (aravindhp) [inline pkg/proxy/envvar.go]: Nit: extra space after `for`.

File: pkg/proxy/envvar_test.go [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"os"
+	"testing"
+
+	"github.com/operator-framework/operator-marketplace/pkg/proxy"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func TestGetOperatorEnvVars(t *testing.T) {
```
Comment (aravindhp) [inline pkg/proxy/envvar_test.go]: Would this test pass even if `TestSetEnvVars()` ran first?

File: pkg/proxy/helpers.go [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	desiredEnvVars := GetOperatorEnvVars()
+
+	if len(desiredEnvVars) != len(actualEnvVars) {
+		return true
+	}
+
+	// Marketplace takes ownership of env vars, and expects them
+	// to be in the order returned by GetOperatorEnvVars.
```
Comment (aravindhp) [inline pkg/proxy/helpers.go]: This is risky. There is no guarantee even then if the ordering will be the same. It is safer to sort and then compare. Take a look at [cache.IsEntryStale()](https://github.com/operator-framework/operator-marketplace/blob/master/pkg/catalogsourceconfig/cache.go#L77).

File: pkg/registry/registry.go [hunk 24]
```diff
@@ -137,6 +138,9 @@ func (r *registry) ensureDeployment(appRegistries []string, needServiceAccount b
 			if needServiceAccount {
 				deployment.Spec.Template.Spec.ServiceAccountName = r.key.Name
 			}
+			// Add, Update, or Remove the proxy env vars based on their values in the marketplace pod.
```
Comment (aravindhp) [inline pkg/registry/registry.go]: I think you are just updating now.

File: pkg/registry/registry.go [hunk 25]
```diff
@@ -354,6 +358,10 @@ func (r *registry) newPodTemplateSpec(registryCommand []string, needServiceAccou
 	if needServiceAccount {
 		podTemplateSpec.Spec.ServiceAccountName = r.key.Name
 	}
+
+	// Add, Update, or Remove the proxy env vars based on their values in the marketplace pod.
```
Comment (aravindhp) [inline pkg/registry/registry.go]: I think you are just updating here.

File: pkg/watches/helpers.go [hunk 26]
```diff
@@ -207,3 +209,54 @@ func getOpSrcOwnerKey(labels map[string]string) *client.ObjectKey {
...
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// If the cluster operator is ever changed we should kick off an event.
+			// Only check if the old name was "cluster", a kube object cannot have it's
+			// name changed in an update.
+			if e.MetaOld.GetName() == proxy.ClusterProxyName {
+				return true
+			}
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
```
Comment (aravindhp) [inline pkg/watches/helpers.go]: You have to take `e.DeleteStateUnknown` into account. Please see the watches for child resources for an example.

File: pkg/apis/operators/v2/catalogsourceconfig_types.go (hunk 18 above)
Comment (awgreene) [reply pkg/apis/operators/v2/catalogsourceconfig_types.go]: Sure.

File: test/testgroups/opsrccsctests.go (hunk 20 above)
Comment (awgreene) [reply test/testgroups/opsrccsctests.go]: That is correct.

File: pkg/proxy/envvar_test.go (hunk 22 above)
Comment (awgreene) [reply pkg/proxy/envvar_test.go]: Good point. I will set the env vars above.

File: pkg/proxy/helpers.go (hunk 23 above)
Comment (awgreene) [reply pkg/proxy/helpers.go]: Understood, I will make sure to sort and then compare the two lists.

File: pkg/proxy/envvar_test.go [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	NO_Proxy    = "NO_PROXY"
+	HTTP_Proxy  = "HTTP_PROXY"
+	HTTPS_Proxy = "HTTPS_PROXY"
+)
+
+func TestGetOperatorEnvVars(t *testing.T) {
```
Comment (aravindhp) [inline pkg/proxy/envvar_test.go]: Nit: I am not sure what feature this test is exercising that is not exercised in `TestSetEnvVars()`

Comment (awgreene) [reply pkg/proxy/envvar_test.go]: That's a good point, should I remove the test?

Comment (awgreene) [reply pkg/proxy/envvar_test.go]: I'll remove the test.

File: cmd/manager/main.go [hunk 28]
```diff
@@ -59,6 +62,17 @@ func main() {
 		log.Fatal(err)
 	}
 
+	// Set proxy API availability
+	k8sInterface, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = proxy.SetProxyAvailability(k8sInterface.Discovery())
```
Comment (ecordell) [inline cmd/manager/main.go]: are we relying on the pod getting restarted if discovery is temporarily unavailable?

**Comment by awgreene** [comment]
Ready for review @ecordell

File: cmd/manager/main.go (hunk 28 above)
Comment (awgreene) [reply cmd/manager/main.go]: Yes - if the discover api fails marketplace would log a fatal error.

File: test/testgroups/opsrccsctests.go [hunk 29]
```diff
@@ -11,6 +11,8 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+var proxyTestsEnalbed = false
```
Comment (ecordell) [inline test/testgroups/opsrccsctests.go]: super-nit: `Enalbed` -> `Enabled`
